### PR TITLE
db setup and initial taxis endpoint

### LINKED ISSUE #1: Config and load information in database
Like the readme [says in HU1](https://github.com/Laboratoria/impl-fleet-management-api-py?tab=readme-ov-file#historia-de-usuario-1-cargar-informaci%C3%B3n-a-base-de-datos), let's use [vercel Postgresql](https://vercel.com/docs/storage/vercel-postgres)

and these files https://drive.google.com/file/d/1T5m6Vzl9hbD75E9fGnjbOiG2UYINSmLx/view

### PR DESCRIPTION
**Pull request by Danieljara23**
## Includes:
- Initial project setup including requirements.txt and gitignore files
- A simple setup for the db connection
- An initial TaxiModel with one single method called get_taxis
- A simple method in app.py to return all the taxis when user navigates to /api/taxis/
- Setup for tests (`client` and `mock_response fixtures`)

## Not included 
- Swagger docs

closes #1

### CODE REVIEW

File: database/db.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+import os
+
+def get_connection():
+    try:
+        return psycopg2.connect(
+          host=os.getenv('POSTGRES_HOST'),
+          user=os.getenv('POSTGRES_USER'),
+          password=os.getenv('POSTGRES_PASSWORD'),
+          database=os.getenv('POSTGRES_DATABASE'))
+    except DatabaseError as ex:
```
Comment (unjust) [inline database/db.py:12]: podemos hacer `psycopg2.Database` o `from psychopg2 import DatabaseError, connect` para evitar 2 imports.

File: models/TaxiModel.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class TaxiModel():
+
+  @classmethod
+  def get_taxis(self):
+    try:
+      connection = get_connection()
+      taxis = []
+
+      with connection.cursor() as cursor:
```
Comment (unjust) [inline models/TaxiModel.py:11]: esta sintaxis `with` es comun? mejor que hacer `cursor = connection.cursor` ?

File: models/TaxiModel.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        cursor.execute("""SELECT  * from taxis""")
+        resultset = cursor.fetchall()
+        
+        for row in resultset:
+            taxi_entry = {
+              'id': row[0],
+              'plate': row[1]
+            }
+            taxis.append(taxi_entry)
```
Comment (unjust) [inline models/TaxiModel.py:20]: me gusta porque es entendible, pero me imagino hay una manera con python en menos lineas.

ChatGPT dice : 
```
taxis = [{'id': row[0], 'plate': row[1]} for row in resultset]
```

File: app.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+app = Flask(__name__)
+    
+@app.get('/api/taxis/')
+def get_taxis():
+  try:     
+    taxis = TaxiModel.get_taxis()
+    response = {
+        'taxis': taxis,
+        'count': len(taxis)
+    }
```
Comment (unjust) [inline app.py:13]: Por ahora creo solo necesitamos una lista (arreglo) de taxis [ { plate, id } ]
Pero el count puede ayudar con la paginacion . Las criterios por este endpoint son
https://github.com/Laboratoria/impl-fleet-management-api-py?tab=readme-ov-file#criterios-de-aceptaci%C3%B3n-1

> El endpoint responde para cada taxi: id y placa.
> El endpoint paginamos los resultados para asegurar que las respuestas sean más fáciles de manejar.

**Review by unjust — COMMENTED** [review]
LGTM , solo pequeños comentarios. seguimos avanzando con paginacion.

File: models/TaxiModel.py (hunk 2 above)
Comment (Danieljara23) [reply models/TaxiModel.py:11]: Yep, el uso de `with` es muy pythonway

File: models/TaxiModel.py (hunk 3 above)
Comment (Danieljara23) [reply models/TaxiModel.py:20]: Buena esa, se llama list comprehension

File: app.py (hunk 4 above)
Comment (Danieljara23) [reply app.py:16]: Done

File: database/db.py (hunk 1 above)
Comment (Danieljara23) [reply database/db.py:11]: Done

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    'plate': 'FGMG-3071',
+  }
+]
+
+
+@pytest.fixture
+def app():
+    """Create application for the tests."""
+    _app = create_app()
+    _app.logger.setLevel(logging.CRITICAL)
```
Comment (unjust) [inline tests/conftest.py:32]: why critical here? maybe we should explain in a comment if its important

File: tests/conftest.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    ctx = _app.test_request_context()
```
Comment (unjust) [inline tests/conftest.py:33]: ok this seems pretty straightforward with the documentation
lets make sure we add 
https://flask.palletsprojects.com/en/3.0.x/testing/#fixtures
https://flask.palletsprojects.com/en/3.0.x/testing/#tests-that-depend-on-an-active-context to resources

File: tests/conftest.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+			},
+			{
+				'id': '10133',
+				'plate': 'PAOF-6727',
+			},
+			{
+				'id': '2210',
+				'plate': 'FGMG-3071',
+			}
+		]
```
Comment (unjust) [inline tests/conftest.py:67]:
```suggestion
		return MOCKED_RESPONSE
```

File: tests/conftest.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import sys
+sys.path.append("..")
+
+import os
+import pytest
+import logging
+
+from models import TaxiModel
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
```
Comment (unjust) [inline tests/conftest.py:9]: sorry, whats this for?

File: tests/conftest.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
+import sys
```
Comment (unjust) [inline tests/conftest.py:1]: the indentation of this file seems weird after line 50

File: tests/conftest.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    ctx.pop()
+
+
+@pytest.fixture
+def client(app):
+    client = app.test_client()
+    yield client
+
+
+class TaxiModelMockResponse:
```
Comment (unjust) [inline tests/conftest.py:50]: maybe just `TaxiModelMock` here?

File: tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+			return {"mock_key": "mock_response"}
+    
+@pytest.fixture
+def mock_response(monkeypatch):
+	"""Requests.get() mocked to return {'mock_key':'mock_response'}."""
+
+	def mock_get(*args, **kwargs):
+			return TaxiModelMockResponse().get_taxis()
+
+	monkeypatch.setattr(TaxiModel.TaxiModel, "get_taxis", mock_get)
```
Comment (unjust) [inline tests/conftest.py:81]: I'm a little confused. Are we using this `mock_response` and `mock_get`? 
If so it just seems like `test_get_taxis` is just confirming what this mock returns, which doesn't make much sense. (Or maybe I don't understand the monkeypatch.)

I think with the client fixture  in the test we want to do a `get` `api/taxis` without a mocked_response, but have the TaxiModel mocked to return the mock data.
That way we will actually be testing our `api/taxis` structures the response right based on the TaxiModel mocked data.

File: tests/test_app.py [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import json
+from .conftest import MOCKED_RESPONSE
+def test_get_taxis(client, mock_response):
+  response = client.get('/api/taxis/')
+  response_data = json.loads(response.get_data(as_text=True))
+  assert response_data['count'] == 3
+  assert response_data['taxis'] == MOCKED_RESPONSE
```
Comment (unjust) [inline tests/test_app.py:7]: Things to test:
* when we call `api/taxis` we get a response that we expect (count, taxis, maybe a header for pagination if we use that)
* when we call with pagination params that TaxiModel.get_taxis is called with the correct params

getting into mocking the cursor and seeing if it sets the offset correctly brings us into more e2e tests, that I think we can wait on.

File: tests/conftest.py (hunk 8 above)
Comment (Danieljara23) [reply tests/conftest.py:9]: This is to let python know we need to import something from a parent folder

File: .gitignore [hunk 13]
```diff
@@ -0,0 +1,2 @@
+venv
+.env
```
[the remark this answers is not in the document]
Comment (Danieljara23) [reply .gitignore:2]: Done

File: app.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        cfg = Config()
+    app = Flask(__name__)
+    app.config.from_object(cfg)
+
+    @app.route("/api/taxis/", methods=["GET"])
+    def get_taxis():
+      try:     
+        page = request.args.get("page", 1, type=int)
+        per_page = request.args.get("per_page", ROWS_PER_PAGE, type=int)
+        taxis = TaxiModel.TaxiModel.get_taxis(page, per_page)
```
[the remark this answers is not in the document]
Comment (Danieljara23) [reply app.py:25]: Done

File: config.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env python
+import os
+
+class Config:
+    DEBUG = True
+    DEVELOPMENT = True
+    SECRET_KEY = "2c084ae6f1b340a7af8880400b83cb49"
+    FLASK_SECRET = SECRET_KEY
```
[the remark this answers is not in the document]
Comment (Danieljara23) [reply config.py:8]: Done. My bad. It's removed now

File: tests/conftest.py (hunk 5 above)
Comment (Danieljara23) [reply tests/conftest.py:32]: Done. Removed

File: tests/conftest.py (hunk 7 above)
Comment (Danieljara23) [reply tests/conftest.py:67]: Done

File: tests/conftest.py (hunk 10 above)
Comment (Danieljara23) [reply tests/conftest.py:50]: Done

**Pull request merged**